Adding metadata tutorial page (#940)
usnistgov/OSCAL: docs/content/learn/tutorials/metadata/_index.md

---
title: Creating and Using Metadata in OSCAL
description: A tutorial that covers the usage of the Metadata Section in OSCAL
weight: 5
suppresstopiclist: true
---

> This tutorial covers the basics of the common Metadata Section, and provides a walk-through on creating one for a new document. Before reading this tutorial you should:

Reviewer: I suggest adding a link to https://pages.nist.gov/OSCAL/concepts/layer/overview/, or at minimum the Metadata definition or a pointer/link to its reference. It is important for people to understand that Metadata is a global assembly found in all OSCAL moddels.

Author: Addressed below.

- Have some familiarity with the [XML](https://www.w3.org/standards/xml/core), [JSON](https://www.json.org/), or [YAML](https://yaml.org/spec/) formats.

> - Have a basic understanding of the usage and organization of the OSCAL Model and it's layers

Author: ```suggestion
- Have a basic understanding of OSCAL Models and their overall structure. ([OSCAL High-Level Overview](https://pages.nist.gov/OSCAL/concepts/layer/overview/))
```

## What is the Metadata Section?

> All OSCAL documents share some common structure and elements, as discussed in [Common High-Level Structure](/documentation/schema/overview/#common-high-level-structure). The foremost of these is the Metadata section, which includes important identifying and categorizing information. A number of mandatory fields inside of Metadata are vital to the processing of OSCAL documents, and have stricter requirements that help ensure interoperability. A second, larger set of fields are optional, and designed to allow OSCAL content creators great flexibility in expressing additional information. These optional fields could enable extra functionality, utilize proprietary data, contain contact info, and more.

Author: ```suggestion
All OSCAL Models share some common structure and elements, as discussed in [Common High-Level Structure](/concepts/layer/overview/#common-high-level-structure). The foremost of these is the Metadata section, which includes important identifying and categorizing information. A number of mandatory fields inside of Metadata are vital to the processing of OSCAL documents, and have stricter requirements that help ensure interoperability. A second, larger set of fields are optional, and designed to allow OSCAL content creators great flexibility in expressing additional information. These optional fields may enable extra functionality, utilize proprietary data, contain contact info, and more.
```

> The definition of the Metadata Section is the same regardless of where it appears, so this tutorial is not bound to a single Layer of OSCAL.

Author: ```suggestion
The definition of the Metadata Section is the same regardless of where it appears, so this tutorial is not bound to a single OSCAL Model.
```

As with all parts of OSCAL, the Metadata Section provides machine-readable formats in XML, JSON, and YAML, which support representing an equivalent set of information. Examples in this tutorial are provided for XML, JSON, and YAML to show the equivalent representations.

## The Fields of Metadata

> The widget below shows the high-level structure of the Metadata Section in XML, JSON, and YAML. Below, each field is defined and briefly explained. Selecting a serialization type here will 

Author: ```suggestion
The widget below shows the high-level structure of the Metadata Section in XML, JSON, and YAML. Below, each field is defined and briefly explained.
```

- [`parties`](https://pages.nist.gov/OSCAL/documentation/schema/catalog-layer/catalog/json-schema/##gen_metadata-party_ref-h3) (Optionally Many) - A responsible entity which is either a person or an organization.
- [`responsible-parties`](https://pages.nist.gov/OSCAL/documentation/schema/catalog-layer/catalog/json-schema/##gen_metadata-responsible-party_ref-h3) (Optionally Many) - A reference to a set of organizations or persons that have responsibility for performing a referenced role in the context of the containing object.
- [`remarks`](https://pages.nist.gov/OSCAL/documentation/schema/catalog-layer/catalog/json-schema/##gen_metadata-remarks_ref-h3) (Optional) - Markup formatted plaintext with notes for human readers of the content.

{{% /tab %}}
{{% /tabs %}}
## Creating a Metadata Section

All OSCAL documents require a Metadata section with a number of basic elements. This tutorial will cover creating this section and populating the required elements. We will also cover several important optional elements, and how they can be used to achieve specialized use cases.

Lets start with a nominal example OSCAL document to explain how you would create Metadata. 

> This OSCAL document is a Catalog, called "Example OSCAL Catalog". It was created today (1/1/2021) by me, Example Person, employee of Example Company.

Author: I feel that both examples are effectively the same already (Example OSCAL Catalog and Sample Security Catalog). Is there value in making sure that they are the same?

> This OSCAL document is a Catalog, called "Example OSCAL Catalog". It was created today (1/1/2021) by me, Example Person, employee of Example Company.

Reviewer: I was hoping we can provide some level of gradual progression and continuity among all tutorials (at least where ever is possible)

> This OSCAL document is a Catalog, called "Example OSCAL Catalog". It was created today (1/1/2021) by me, Example Person, employee of Example Company.

Reviewer 2: While I think this might be helpful, some decoupling may be needed. Maintaining too much connectivity between examples in the tutorials is going to make these hard to maintain as we make changes.

{{% /tab %}}
{{% tab %}}
{{< highlight yaml "linenos=table" >}}
---

catalog:
  uuid: c3da6d1d-c20c-4c7c-ae73-4010167a186

{{< /highlight >}}

Although technically outside of the Metadata Section, the `uuid` of a document is always given as a field of the root element, and shares its structure across all OSCAL document types. This is provided using the [uuid](https://pages.nist.gov/OSCAL/documentation/schema/datatypes/#uuid) datatype. UUIDs are a globally unique identifier randomly generated at the same time an OSCAL document is created. This provides a stable and unique way to refer to a given instance of an OSCAL document. Many tools provide easy ways to generate version 4 UUIDs. For our example we have generated one using a trivial UUIDv4 generator.

{{% /tab %}}
{{% /tabs %}}

## Starting Simple with Required Elements

{{< tabs XML JSON YAML >}}

{{% tab %}}

First, we start with the `<title>` element. This should be populated with a human-readable title that is exactly or closely matches the existing name of the document. 

For our example, this would look like:

{{< highlight xml "linenos=table" >}}

<title> Example OSCAL Catalog </title>

{{< /highlight >}}

> Next we have two date based elements. This uses the OSCAL DataTime data type as defined here: [date-with-timezone](https://pages.nist.gov/OSCAL/documentation/schema/datatypes/#date-with-timezone). `<published>` gives the DateTime of when the document was published for the first time. Since we are publishing this document for the first time today (1/1/2021), the published date will be today, and the time is the moment (or near to it) that this document was published. `<last-modified>` provides the DateTime of the most recent change to this document. As we have never published this particular document before, the DateTime should be identical to the one given by `<published>`. In our example, this will look like:

Author: ```suggestion
Next we have two date-based elements. This uses the OSCAL DataTime data type as defined here: [date-with-timezone](https://pages.nist.gov/OSCAL/documentation/schema/datatypes/#date-with-timezone). `<published>` gives the DateTime of when the document was published for the first time. Since we are publishing this document for the first time today (1/1/2021), the published date will be today, and the time is the moment (or near to it) that this document was published. `<last-modified>` provides the DateTime of the most recent change to this document. As we have never published this particular document before, the DateTime should be identical to the one given by `<published>`. In our example, this will look like:
```

{{< highlight yaml "linenos=table" >}}
---
  version: 1.0.0
{{< /highlight >}}

Where possible, use well formatted versions with clear and defined rules. This version will be incremented whenever the OSCAL document is updated.

The final required field is `oscal-version`, which provides consumers of the document that the version of OSCAL that was used to create this document.

{{< highlight yaml "linenos=table" >}}
---
{
oscal-version: 1.0.0
}
{{< /highlight >}}

We have now covered all required fields of the Metadata Section, and could now publish our document as a valid OSCAL document. Lets take a look at what it would look like all together:

{{< highlight yaml "linenos=table" >}}
---
  catalog:
    uuid: c3da6d1d-c20c-4c7c-ae73-4010167a186,
    metadata:
      title: Example OSCAL Catalog,
      published: 2021-01-01T00:00:00-5:00,
      last-modified: 2021-01-01T00:00:00-5:00,
      version: 1.0.0,
      oscal-version: 1.0.0
{{< /highlight >}}

However, the Metadata Section has several other optional fields of importance, which we will cover in the following sections.
{{% /tab %}}
{{% /tabs %}}

> ## Understanding and using `document-id`

Author: ```suggestion
## Understanding and Using `document-id`
```

{{< highlight yaml "linenos=table" >}}
---
document-ids:
  - scheme: https://www.doi.org/
    value: 10.1000/182
{{< /highlight >}}

In the future, any updated versions of this document we publish will have the same `document-id` value, but a different `uuid`.

In the case that a document does not explicitly provide a `document-id` value, it is given one implicitly that is equal to it's `uuid`. This means that if we publish a new document without a `document-id`, then later need to publish an update to it, we can tie it back to the original document.
{{% /tab %}}
{{% /tabs %}}

## Using Props
{{< tabs XML JSON YAML >}}
{{% tab %}}
The `<prop>` element provides a key-value pairing to provide additional information alongside the well-defined elements of the Metadata Section. Some common use cases include labels, sorting tags, and classification. There are no restrictions on using custom `name` and `value` pairs making `<prop>` an excellent place to do extension of the OSCAL model to meet proprietary use cases. The Metadata Section may include any number of `<prop>` elements.

> For our example, we will create a `<prop>` to mark an OSCAL document with a [Traffic Light Protocol (TLP)](https://en.wikipedia.org/wiki/Traffic_Light_Protocol) classification. 

Author: Would be happy to change the example used here if we can write one up.

> For our example, we will create a `<prop>` to mark an OSCAL document with a [Traffic Light Protocol (TLP)](https://en.wikipedia.org/wiki/Traffic_Light_Protocol) classification. 

Reviewer: Please let me know if you need help with creating one example.

  - successor-version: This link identifies a resource containing the predecessor version in the version history. RFC 5829.

For this example, we will create a "latest-version" link, which provides a static link to the latest version of this Document Series. By pointing to a static URL, but updating the contents of that URL as the document is updated, any user that has an old version of the document can quickly and easily update.

{{< highlight xml "linenos=table" >}}
<link rel="latest-version" href="https://www.examplecompany.com/catalog/exampleoscalcatalog/latest"/>
{{< /highlight >}}

The "predecessor-version" and "successor-version" links could be used in tandem with the above to create a navigable web of document versions.

By using other custom `rel` values, any relevant or related resource can be described and linked to in the OSCAL Document's Metadata Section.

Use of the optional `media-type` attribute can provide consumers with processing instructions based on the media type of the resource found at the `href`. See [IANA Media Types](https://www.iana.org/assignments/media-types/media-types.xhtml).
{{% /tab %}}

{{% tab %}}

> Alongside the `"props"` array, the `"links"` array is way of providing additional information in the Metadata section that is not covered by the other fields. Each object inside the `"links"` array must contain a single `"href"` field, which contains a resolvable URI. Each object should also include a `"rel"` field, which describes the type of relationship provided by the `"href"`. Although any string can be used in the `"rel"` field, OSCAL identifies five common link relationships that are frequently useful:

Author: ```suggestion
Alongside the `"props"` array, the `"links"` array is a way of providing additional information in the Metadata section that is not covered by the other fields. Each object inside the `"links"` array must contain a single `"href"` field, which contains a resolvable URI. Each object should also include a `"rel"` field, which describes the type of relationship provided by the `"href"`. Although any string can be used in the `"rel"` field, OSCAL identifies five common link relationships that are frequently useful:
```

Will address capitals in separate commit.
